File: src/java/org/apache/fop/afp/util/StringUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.afp.util;

/**
 * Library of utility methods useful in dealing with strings.
 *
 */
public final class StringUtils {

    private StringUtils() {
    }

    /**
     * Padds the string to the left with the given character for
     * the specified length.
     * @param input The input string.
     * @param padding The char used for padding.
     * @param length The length of the new string.
     * @return The padded string.
     */
    public static String lpad(String input, char padding, int length) {

        if (input == null) {
            input = new String();
        }

        if (input.length() >= length) {
            return input;
        } else {
            StringBuffer result = new StringBuffer();
            int numChars = length - input.length();
            for (int i = 0; i < numChars; i++) {
                result.append(padding);
            }
            result.append(input);
            return result.toString();
        }
    }

    /**
     * Padds the string to the right with the given character for
     * the specified length.
     * @param input The input string.
     * @param padding The char used for padding.
     * @param length The length of the new string.
     * @return The padded string.
     */
    public static String rpad(String input, char padding, int length) {

        if (input == null) {
            input = new String();
        }

        if (input.length() >= length) {
            return input;
        } else {
            StringBuffer result = new StringBuffer(input);
            int numChars = length - input.length();
            for (int i = 0; i < numChars; i++) {
                result.append(padding);
            }
            return result.toString();
        }
    }

}

File: dotnet/src/Fop.Util/StringUtils.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace Fop.Util;

/// <summary>
/// String helper utilities.
/// <para>Port of <c>org.apache.fop.util.StringUtils</c>.</para>
/// </summary>
public static class StringUtils
{
    /// <summary>
    /// Replaces every soft hyphen (U+00AD) with an ordinary hyphen-minus ('-') unless the
    /// supplied painter declares support for soft hyphens, in which case the text is returned
    /// unchanged.
    /// </summary>
    /// <param name="text">the text to process.</param>
    /// <param name="painter">the painter whose capability decides the replacement.</param>
    /// <returns>the processed text.</returns>
    public static string ProcessSoftHyphen(string text, ISoftHyphenSupport painter)
    {
        ArgumentNullException.ThrowIfNull(text);
        ArgumentNullException.ThrowIfNull(painter);
        return painter.SupportsSoftHyphen
            ? text
            : text.Replace(CharUtilities.SoftHyphen, '-');
    }
}

/// <summary>
/// Abstraction of the single capability of <c>org.apache.fop.render.intermediate.IFPainter</c>
/// that <see cref="StringUtils.ProcessSoftHyphen"/> depends on.
/// </summary>
/// <remarks>
/// TODO: when the intermediate-format painter (<c>IFPainter</c>) is ported, have it implement
/// this interface (or fold this capability into it) so callers pass the real painter directly.
/// </remarks>
public interface ISoftHyphenSupport
{
    /// <summary>Gets a value indicating whether the painter supports soft hyphens.</summary>
    bool SupportsSoftHyphen { get; }
}
